### DISCUSSION TITLE
keyword as List

### DISCUSSION THREAD

**Discussion by dongri727** [serverpod/serverpod#1303] [Q&A] [answered]
Can the keyword of the get function be a List? I want to do a search that connects multiple keywords with OR, and ChatGPT recommends the following code, but it doesn't work. please tell me how to write it to enable OR search.

/*   Future<List<Principal>> getPrincipal(Session session,
      {String? keyword}) async {
    print("Getting principal with keyword: $keyword");
    return await Principal.find(
      session,
      where: (t) =>
          keyword != null ? t.pays.like('%$keyword%') : Constant(true),
      orderBy: Principal.t.annee,
    );
  } */

  Future<List<Principal>> getPrincipal(Session session,
    {List<String>? keywords}) async {
  print("Getting principal with keywords: $keywords");

  var whereClause;
  if (keywords != null && keywords.isNotEmpty) {
    for (var keyword in keywords) {
      if (whereClause == null) {
        whereClause = Principal.t.pays.like('%$keyword%');
      } else {
        whereClause = whereClause | Principal.t.pays.like('%$keyword%');
      }
    }
  } else {
    whereClause = Constant(true);
  }

  return await Principal.find(
    session,
    where: whereClause,
    orderBy: Principal.t.annee,
  );
}

**Comment by SandPod** [comment]
Hey @dongri727 👋 

Expressions support having multiple `OR` clauses.
As a first step in your troubleshooting, It seems like you have missed defining what type the List should contain.

Try changing:
```dart
Future<List> getPrincipal(Session session, {List? keywords}) async {
    print("Getting principal with keywords: $keywords");

    var whereClause;
    if (keywords != null && keywords.isNotEmpty) {
        for (var keyword in keywords) {
            if (whereClause == null) {
                whereClause = Principal.t.pays.like('%$keyword%');
            } else {
                whereClause = whereClause | Principal.t.pays.like('%$keyword%');
            }
        }
    } else {
        whereClause = Constant(true);
    }

    return await Principal.find(
        session,
        where: whereClause,
        orderBy: Principal.t.annee,
    );
}
```

To 
``` dart
Future<List<Principal>> getPrincipal(Session session, {List<String>? keywords}) async {
    print("Getting principal with keywords: $keywords");

    var whereClause;
    if (keywords != null && keywords.isNotEmpty) {
        for (var keyword in keywords) {
            if (whereClause == null) {
                whereClause = Principal.t.pays.like('%$keyword%');
            } else {
                whereClause = whereClause | Principal.t.pays.like('%$keyword%');
            }
        }
    } else {
        whereClause = Constant(true);
    }

    return await Principal.find(
        session,
        where: whereClause,
        orderBy: Principal.t.annee,
    );
}
```

It is also OK to pass in a null whereClause so your statement for `whereClause = Constant(true);` could be removed.

**Comment by dongri727** [comment]
Thank you for your response.
I made the changes you suggested and ran the program, but the Flutter console shows

flutter: call failed: principal.getPrincipal
flutter: ServerpodClientException: Internal server error. Call log ID: 14, statusCode = 500

Terminal log.

Get principal by keyword: [Universe, Milky Way].
METHOD CALL: principal.getPrincipal duration: 11ms numQueries: 0 authenticatedUser: null
Type 'Expression' is not a subtype of type '((PrincipalTable) => Expression)? is not a subtype of type '((PrincipalTable) => Expression)?
#0 PrincipalEndpoint.getPrincipal (package:acorn_server/src/endpoints/principal_endpoint.dart:29:12)
#1 Endpoints.initializeEndpoints.<anonymous closure> (package:acorn_server/src/generated/endpoints.dart:1182:66)
#2 EndpointDispatch.handleUriCall (package:serverpod/src/server/endpoint_dispatch.dart:111:37)
<asynchronous suspend
#3 Server._handleRequest (package:serverpod/src/server/server.dart:243:18)
<asynchronous suspend

It is displayed and cannot be retrieved.
Any guidance on how to fix this would be appreciated.　Thank You

**Reply by SandPod** [reply, marked as answer]
Then I think the issues Is in the error message.

If you look at the method signature for `Principle.find` you will se that the type of the `where` parameter is a `PrincipleExpressionBuilder` which is a function that takes a `PrincipleTable` and returns an `Expression`; `((PrincipalTable) => Expression)?`.

If I understand your code correctly you have already defined the Expression you would like to return as `whereClause`. Now you just need to make sure it is returned from the `PrincipleExpressionBuilder` lambda.

Try this:
``` dart
Future<List<Principal>> getPrincipal(Session session, {List<String>? keywords}) async {
    print("Getting principal with keywords: $keywords");

    var whereClause;
    if (keywords != null && keywords.isNotEmpty) {
        for (var keyword in keywords) {
            if (whereClause == null) {
                whereClause = Principal.t.pays.like('%$keyword%');
            } else {
                whereClause = whereClause | Principal.t.pays.like('%$keyword%');
            }
        }
    } else {
        whereClause = Constant(true);
    }

    return await Principal.find(
        session,
        where: (_) => whereClause,
        orderBy: Principal.t.annee,
    );
}
```

**Comment by dongri727** [comment]
Thank you very much, I'll try it　　and it works ! thank you !
